vpr/src/pack/pack.cpp

```diff
@@ -156,30 +155,45 @@ bool try_pack(t_packer_opts* packer_opts,
             clustering_data);
 
         //Try to size/find a device
-        bool fits_on_device = try_size_device_grid(*arch, helper_ctx.num_used_type_instances, packer_opts->target_device_utilization, packer_opts->device_layout);
+        std::vector<std::string> overused_blocks = try_size_device_grid(*arch, helper_ctx.num_used_type_instances, packer_opts->target_device_utilization, packer_opts->device_layout);
 
         /* We use this bool to determine the cause for the clustering not being dense enough. If the clustering
          * is not dense enough and there are floorplan constraints, it is presumed that the constraints are the cause
          * of the floorplan not fitting, so attraction groups are turned on for later iterations.
          */
         bool floorplan_not_fitting = (floorplan_regions_overfull || g_vpr_ctx.mutable_floorplanning().constraints.get_num_partitions() > 0);
 
-        if (fits_on_device && !floorplan_regions_overfull) {
+        if (overused_blocks.empty() && !floorplan_regions_overfull) {
             break; //Done
         } else if (pack_iteration == 1 && !floorplan_not_fitting) {
             //1st pack attempt was unsucessful (i.e. not dense enough) and we have control of unrelated clustering
             //
             //Turn it on to increase packing density
```

Comment:
This is a long subroutine, so it would be better to pull this auto-packing code all new code + some of the packer_opts->balance_block_type_utilization code) into a new helper function.